### PR TITLE
Improved collector, revert bson float support, add streaming and better dynamic collectors

### PR DESCRIPTION
**Pull request by tychoish**

### CODE REVIEW

File: read.go [hunk 1]
```diff
@@ -134,12 +120,12 @@ func readChunks(ctx context.Context, ch <-chan *bson.Document, o chan<- Chunk) e
...
 				metrics[i].Values[j] = delta
 			}
-			metrics[i].Values = append([]int64{v.startingValue}, undelta(v.startingValue, metrics[i])...)
+			metrics[i].Values = undelta(v.startingValue, metrics[i].Values)
 		}
 		select {
 		case o <- Chunk{
 			metrics:   metrics,
-			nPoints:   ndeltas + 1,
+			nPoints:   ndeltas + 1, // this accounts for the reference document
```
[the remark this answers is not in the document]
Comment (tychoish) [reply read.go]: yea

File: ftdc_test.go [hunk 2]
```diff
@@ -207,3 +208,113 @@ func TestRoundtrip(t *testing.T) {
...
+		} {
+			t.Run(test.name, func(t *testing.T) {
+				out := delta(test.previous, test.sample)
+				assert.Equal(t, test.expected, out)
+			})
+		}
+	})
+	t.Run("RoundTrip", func(t *testing.T) {
+		t.Run("Empty", func(t *testing.T) {
+			t.Skip("rt broken")
```
[the remark this answers is not in the document]
Comment (tychoish) [reply ftdc_test.go]: I took out these tests, as delta function isn't being tested like this so the model probably doesn't make a lot of sense.

File: collector_better_test.go [hunk 3]
```diff
@@ -16,122 +16,214 @@ func TestEncoding(t *testing.T) {
...
-		},
-		{
-			name:    "SmallDecreasesStall",
-			dataset: []int64{10, 9, 9, 9, 9},
-		},
-		{
-			name:    "SmallRandSomeNegatives",
-			dataset: []int64{rand.Int63n(100), -1 * rand.Int63n(100), rand.Int63n(100), -1 * rand.Int63n(100)},
+			name:    "Better",
+			factory: func() Collector { return &betterCollector{maxDeltas: 20} },
```
[the remark this answers is not in the document]
Comment (tychoish) [reply collector_better_test.go]: Done

File: collector_better.go [hunk 4]
```diff
@@ -107,44 +118,41 @@ func (c *betterCollector) getPayload() ([]byte, error) {
...
+				continue
+			}
+
+			if zeroCount > 0 {
+				payload.Write(encodeValue(0))
+				payload.Write(encodeValue(zeroCount - 1))
+				zeroCount = 0
+			}
+
+			payload.Write(encodeValue(delta))
```
[the remark this answers is not in the document]
Comment (tychoish) [reply collector_better.go]: when delta == 0 you skip it and just increment a zeroRun counter, and then when you get to the first non-zero delta, you write out zeros (plus the count,) and then you write the delta.

[the remark this answers is not in the document]
Comment (julianedwards) [reply collector_better.go]: ah yup, i see now

**Pull request merged**